---
output: pdf_document
geometry: margin=1in
---
  
```{r,fig.width=7.5, fig.height=5,warning=F, message=F, echo=F,results='asis'}

library(dplyr)
library(ggplot2)
library(data.table)
library(grid)
library(stringr)
library(gridExtra)
library(knitr)
library(tidyr)
library(scales)
setwd("/Users/kristen/Documents/transposon_figure_data/data")
load("Processed_Transposon_Mappings.Rda")
load("position_QTL.Rda")
load("count_QTL.Rda")
#load("away_phenos.Rda")
#load("median_phenos.Rda")
#load("low_ld.Rda")
#load("counts_to_remove.Rda")
#load("Amedian.Rda")





#remove fraction and movement traits
#processed_mapping_df<-subset(processed_mapping_df,
#                                grepl('^I', processed_mapping_df$trait) |
#                                 grepl('^V', processed_mapping_df$trait) |
#                                grepl('^X', processed_mapping_df$trait)|
#                               grepl('_C$', processed_mapping_df$trait))
#processed_mapping_df<-subset(processed_mapping_df,!grepl('^no_', processed_mapping_df$trait))


# make cure C and frac are not getting confused
#PxG function
hm<-processed_mapping_df
hm<-distinct(processed_mapping_df, trait,strain,peak_id)
hm$allele <- factor(hm$allele,
                    levels = c(-1,1),
                    labels = c("Ref", "Alt"))
c<-filter(processed_mapping_df, log10p> BF)
#d<-filter(processed_mapping_df,peak_id=="2", trait=="absent_TRANS_CER1_C")
e<-filter(processed_mapping_df,!is.na(peak_id), trait=="absent_TRANS_CER1_C")
f<-distinct(e,POS,peak_id)

gwasPxG <- function(trt,specific_peak){
  #load("~/Dropbox/AndersenLab/RCode/Stefan/good_gwasMappingsINlinkage_phenotypes.Rda")
  
  hm %>%
    filter(trait==trt,peak_id==specific_peak)%>%
    ggplot(.)+
    aes(x=allele,y = value,fill=as.factor(allele))+
    geom_point(size = 1, alpha = .8,position=position_jitter(w=.4,  h=.025),na.rm=TRUE)+
    geom_boxplot(outlier.shape=NA,size =1,color="hotpink2")+
    #scale_fill_brewer(palette = "Set2")+
    
    #geom_point(size = 1, alpha = .8,position=position_jitter(w=.4,  h=.025),na.rm=TRUE)+
    # geom_jitter(size = 3, alpha = .8,postion=position_jitter(0,  0))  +
    theme_bw()+
    theme(axis.text.x = element_text(size=9, , color="black"),
          axis.text.y = element_text(size=9,  color="black"),
          axis.title.x = element_text(size=9,  color="black"),
          axis.title.y = element_text(size=9,  color="black",vjust=1),
          strip.text.x = element_text(size=9,  color="black"),
          strip.text.y = element_text(size=9,  color="black"),
          plot.title = element_text(size=9,  vjust=1),
          legend.title = element_text(size=9),
          panel.border = element_rect(size=1, colour = "black"),
          plot.margin = unit(c(.05,.05,.05,.05), "cm"),
          legend.position = "none")+
    scale_y_continuous(breaks= pretty_breaks())+
    labs( x = "Genotype",y="Value")+
    
    scale_fill_manual( values = c("steelblue3", "indianred1", "darkolivegreen","black"))
}

# pull unique combos, remove strain column(don't need specific strain info at this point)
#processed_mapping_df <- distinct(select(processed_mapping_df, -strain,-allele,-value))
processed_mapping_df<- processed_mapping_df %>% distinct(trait,marker,strain)


#create family and method columns
processed_mapping_df$family <- stringr::str_split_fixed(processed_mapping_df$trait, "_TRANS_",2)[,2]
processed_mapping_df$method <- stringr::str_split_fixed(processed_mapping_df$trait, "_TRANS_",2)[,1]


processed_mapping_df<-filter(processed_mapping_df, family=="CER1_C")

#set up method labellers
method_names <- list(
  'absent'="absence",
  'new'="insertion",
  'reference'="reference"
)

method_labeller <- function(variable,value){
  if (variable=='method') {
    return(method_names[value])
  }else {
    return(as.character(value))
  }
}


#read in position data and create family column
positions <- read.table("CtCp_all_nonredundant.txt",header=TRUE)
names(positions)<-c("CHROM","start","end","TE","orientation","method","strain","class")
positions$family<- stringr::str_split_fixed(positions$TE, regex("_(non-)?reference"),2)[,1]
positions$family<- paste(stringr::str_split_fixed(positions$family, "_",4)[,3],stringr::str_split_fixed(positions$family, "_",4)[,4],sep="_")
positions$family <- gsub("_$" ,"",positions$family)
positions$family <- gsub("_non-reference(.*)$" ,"",positions$family)

#select traits above BF.....this step not needed, double checking everything is above BF
selection<-filter(processed_mapping_df, log10p > BF)

#extract the count base traits
base_traits <-selection[(selection$method=="absent"| selection$method=="new" |selection$method=="reference"|selection$method=="ZERO_new"|selection$method=="ONE_new"), ]
counts<-subset(base_traits, grepl("_C$", base_traits$family))
counts$family <- gsub("_C$" ,"",counts$family)

#
#
###REMOVE LATER!!!!!



#
#
processed_mapping_df <- distinct(select(processed_mapping_df, -strain,-allele,-value))
processed_mapping_df<- processed_mapping_df %>% distinct(trait,marker)



#pull out only position traits from mappings dataframe
position_traits<-subset(selection,
                        grepl('^I', selection$trait) |
                          grepl('^V', selection$trait) |
                          grepl('^X', selection$trait))

#create family column
position_traits$family  <- paste(stringr::str_split_fixed(position_traits$trait, "_",4)[,3],stringr::str_split_fixed(position_traits$trait, "_",4)[,4],sep="_")
position_traits$family <- gsub("_$" ,"",position_traits$family)
position_traits$family <- gsub("_non-reference(.*)$" ,"",position_traits$family)

#optional filter for away phenos
#position_traits<-position_traits[position_traits$trait %in% away$trait,] 
#optional filter for different median phenos 
#position_traits<-position_traits[position_traits$trait %in% median_df$trait,] 
#optional filter for low ld phenos
#position_traits<-position_traits[position_traits$trait %in% low_ld$trait,] 

# add position TRAIT_col family info to processed_mapping_df
processed_mapping_df<-processed_mapping_df %>%mutate(family = ifelse(processed_mapping_df$trait %in% position_traits$trait, (paste(stringr::str_split_fixed(processed_mapping_df$trait, "_",4)[,3],stringr::str_split_fixed(processed_mapping_df$trait, "_",4)[,4],sep="_")), processed_mapping_df$family))


#bind count and position traits option...choose oen of below two
selection<-rbind(counts,position_traits)
#selection<-counts

#COMMENT OUT BELOW!!!!!!

#selection<-selection[selection$trait=="ONE_new_TRANS_Tc1A_C"|selection$trait=="absent_TRANS_MIRAGE1_C",]


#selection<-selection[selection$trait=="I_13455974_Tc3"|selection$trait=="V_739436_WBTransposon00000074",]
#selection<-selection[selection$trait=="II_507265_LINE2C",]
#selection<-selection[selection$trait=="II_507265_LINE2C",]
#absent_TRANS_CER1

selection<-selection[selection$trait=="absent_TRANS_CER1_C",]


#

#strip count marker and remnant marks from dataframes
selection$trait <- gsub("_C$" ,"",selection$trait)
hm$trait <- gsub("_C$" ,"",hm$trait)
processed_mapping_df$trait <- gsub("_C$" ,"",processed_mapping_df$trait)
processed_mapping_df$family <- gsub("_C$" ,"",processed_mapping_df$family)
processed_mapping_df$family <- gsub("_$" ,"",processed_mapping_df$family)
processed_mapping_df$family <- gsub("_non-reference(.*)$" ,"",processed_mapping_df$family)

processed_mapping_df<-mutate(processed_mapping_df,ID=paste(trait,peak_id,sep="_"))
copy<-processed_mapping_df
processed_mapping_df<-mutate(processed_mapping_df,SNP_col=ifelse(ID %in% position_QTL$trait | ID %in% count_QTL$trait,"PASS","FAIL" ))
#d<-filter(processed_mapping_df,ID=="absent_TRANS_CER1_3")
#d<-filter(processed_mapping_df,family=="CER1")
#### NEED TO FIX HERE

count_QTL<-mutate(count_QTL, trait2=gsub("_\\d+$","",trait)) 
position_QTL<-mutate(position_QTL, trait2=gsub("_\\d+$","",trait))  
selection <- filter(selection, (trait %in% count_QTL$trait2 |trait %in% position_QTL$trait2))
#unique(count_QTL$trait2)
#unique(position_QTL$trait2)
#unique(selection$trait)
processed_mapping_df<-filter(processed_mapping_df,CHROM != "MtDNA")




i="absent_TRANS_CER1"


#iterate through the phenotypes and plot the results
for (i in unique(selection$trait)){
  specific_trait<- processed_mapping_df[processed_mapping_df$trait == i, ]
  empty <-specific_trait[specific_trait$method==NA,]
  #specific_trait_mx <- max(specific_trait$log10p)
  pvalues<-filter(specific_trait,log10p !="Inf") #
  specific_trait_mx <- max(pvalues$log10p) #
  TE<-specific_trait$family[1]
  rect_data<-filter(specific_trait,SNP_col=="PASS")
  ##check for NAs
  #sapply(Mappings, function(x)all(is.na(x)))
  A<- processed_mapping_df %>%
    filter(trait == i)%>%
    ggplot(.)+
    aes(x=POS/1e6,y=log10p)+
    geom_rect( data=rect_data,mapping=aes(xmin=startPOS/1e6, xmax=endPOS/1e6, ymin=0, ymax= specific_trait_mx),fill="thistle1", alpha=1)+
    geom_point(aes( color=ifelse(log10p> BF & SNP_col=="PASS", 'red', 'black')),size=1)+
    
    facet_grid(.~CHROM,scale="free_x",space = "free_x")+scale_color_identity()+
    ggtitle(i)+
    geom_hline(aes(yintercept=BF),color="grey60",linetype="dashed")+
    theme(strip.background = element_rect(fill = "white"),
          strip.text.x = element_text(size = 9, colour = "black",face="bold"),
          panel.background = element_rect(fill = "white"),
          panel.border = element_rect(color="black", size=0.5, linetype="solid", fill=NA),
          panel.margin = unit(.6, "lines"),
          axis.ticks =element_line(colour = "black"),
          axis.text.y = element_text(colour = "black"),
          axis.text.x = element_text(colour = "black"),
          axis.title=element_text(size=9),
          plot.margin=unit(c(.1,.1,-.25,.1), "cm"),
          legend.position=('none'))+
    labs(x="",y="-log10(p)")
  
  #scale_y_continuous(expand=c(0,0),limits=c(0,specific_trait_mx+.075*specific_trait_mx),labels = function(x) format(x,width = 4))

  # pull out  X maxs of each panel
  panel1<-filter(ggplot_build(A)$data[[2]],PANEL==1)
  max1<-max(panel1$x)
  min1<-(min(panel1$x))
  panel2<-filter(ggplot_build(A)$data[[2]],PANEL==2)
  max2<-(max(panel2$x))
  min2<-(min(panel2$x))
  panel3<-filter(ggplot_build(A)$data[[2]],PANEL==3)
  max3<-(max(panel3$x))
  min3<-(min(panel3$x))
  panel4<-filter(ggplot_build(A)$data[[2]],PANEL==4)
  max4<-(max(panel4$x))
  min4<-(min(panel4$x))
  panel5<-filter(ggplot_build(A)$data[[2]],PANEL==5)
  max5<-(max(panel5$x))
  min5<-(min(panel5$x))
  panel6<-filter(ggplot_build(A)$data[[2]],PANEL==6)
  max6<-(max(panel6$x))
  min6<-(min(panel6$x))
  
  positions$TRAIT_col<-paste(positions$method, "TRANS", positions$family, sep="_")
  #traitPositions<-positions[positions$TRAIT_col==i,]
  traitPositions<-positions[positions$family==TE,]
  
  blank <- data.frame(CHROM=character(),
                      start=integer(),
                      end=integer(),
                      TE=character(),
                      orientation=character(),
                      method=character(),
                      strain=character(),
                      class=character(),
                      family=character(),
                      TRAIT_col=character(),
                      stringsAsFactors=FALSE)
  
  blank[1,]<- c( "I", as.integer(10000000),as.integer(10000000),"blank","+","blank","fake","blank","blank","blank")
  blank[2,]<- c( "II", as.integer(10000000),as.integer(10000000),"blank","+","blank","fake","blank","blank","blank")
  blank[3,]<- c( "III", as.integer(10000000),as.integer(10000000),"blank","+","blank","fake","blank","blank","blank")
  blank[4,]<- c( "IV", as.integer(10000000),as.integer(10000000),"blank","+","blank","fake","blank","blank","blank")
  blank[5,]<- c( "V", as.integer(10000000),as.integer(10000000),"blank","+","blank","fake","blank","blank","blank")
  blank[6,]<- c( "X", as.integer(10000000),as.integer(10000000),"blank","+","blank","fake","blank","blank","blank")
  
  traitPositions<-rbind(traitPositions,blank)
  traitPositions$start<-as.integer(traitPositions$start)
  #m <- ggplot(summarydata, aes(x=start/1e6,fill=class))
  #m <-m + geom_histogram(binwidth=.25)+
  
  #ggplot(data = combo, aes(x = TEMP_support,y=TELOCATE_support,color=ifelse(method=="absent","darkorange",ifelse(method=="blank","black",ifelse(method=="insertion",""turquoise3"","slateblue1")))))+scale_color_identity()
  #    m <- ggplot(traitPositions, aes(x=start/1e6,color=ifelse(method=="absent","darkorange",ifelse(method=="blank","black",ifelse(method=="new","turquoise3","slateblue1")))))+scale_color_identity()
  m <- ggplot(traitPositions, aes(x=start/1e6,fill=method))+scale_fill_manual(values = c('new' = "turquoise3", 'reference' = "slateblue1", 'absent' = "darkorange","blank"="black"))
  m <-m + geom_histogram(data=subset(traitPositions,strain=="fake"), fill="white", colour="white", binwidth=.25)
  m <-m + geom_histogram(data=subset(traitPositions,strain!="fake"), binwidth=.25)+
    facet_grid(. ~ CHROM,scale="free",drop=FALSE)+
    ggtitle("")+
    geom_point(data = subset(traitPositions, CHROM=="I"),aes(x=max1,y=0),alpha=0)+
    geom_point(data = subset(traitPositions, CHROM=="II"),aes(x=max2,y=0),alpha=0) +
    geom_point(data = subset(traitPositions, CHROM=="III"),aes(x=max3,y=0),alpha=0) +
    geom_point(data = subset(traitPositions, CHROM=="IV"),aes(x=max4,y=0),alpha=0) +
    geom_point(data = subset(traitPositions, CHROM=="V"),aes(x=max5,y=0),alpha=0) +
    geom_point(data = subset(traitPositions, CHROM=="X"),aes(x=max6,y=0),alpha=0) +
    
    geom_point(data = subset(traitPositions, CHROM=="I"),aes(x=min1,y=0),alpha=0) +
    geom_point(data = subset(traitPositions, CHROM=="II"),aes(x=min2,y=0),alpha=0) +
    geom_point(data = subset(traitPositions, CHROM=="III"),aes(x=min3,y=0),alpha=0) +
    geom_point(data = subset(traitPositions, CHROM=="IV"),aes(x=min4,y=0),alpha=0) +
    geom_point(data = subset(traitPositions, CHROM=="V"),aes(x=min5,y=0),alpha=0) +
    geom_point(data = subset(traitPositions, CHROM=="X"),aes(x=min6,y=0),alpha=0)+
    
    labs(x = "Chromosome Position (Mb)", y="Number of Transposition Events")+
    theme(strip.background = element_blank(),
          strip.text.x = element_blank(),
          #strip.text = element_text(size = 9, colour = "black",face="bold"),
          panel.margin = unit(.6, "lines"),
          panel.border = element_rect(fill=NA,colour = "black"),
          panel.background = element_rect(fill = "white"),
          axis.ticks =element_line(colour = "black"),
          axis.title=element_text(size=9),
          axis.text.y = element_text(colour = "black",size=9),
          axis.text.x = element_text(colour = "black",size=9),
          legend.title=element_blank(),
          # legend.position="bottom",
          plot.margin=unit(c(-.25,.1,.1,.1), "cm"),
          legend.position=('none'))
  m
  
  #now can check plot for max value and set y limit to a certain percent above that max value 
  m <- m + scale_y_continuous(expand = c(0,0),limits=c(0,max(ggplot_build(m)$panel$ranges[[1]]$y.range)*1.075)) 
  
  
  library(gtable)
  g1<-ggplotGrob(A)
  g2<-ggplotGrob(m)
  #Bind the tables
  g<-gtable:::rbind_gtable(g1, g2, "first")
  #Remove a row between the plots
  #g <- gtable_add_rows(g, unit(-1,"cm"), POS=nrow(g1))
  #draw
  panels <- g$layout$t[grep("panel", g$layout$name)]
  g$heights[panels] <- lapply(c(10,10), unit, "null")
  grid.newpage()
  grid.draw(g)
  
  
  df <- data.frame(1,2)
  blank_plot<-ggplot(df,aes(x=1,y=1)) + geom_point(color="white") +
    theme(axis.text =element_blank(),axis.ticks =element_blank(),axis.title =element_blank(),panel.background = element_blank(),panel.grid = element_blank())
  
  plist<-vector()
  for (p in rect_data$peak_id){
    plist<-c(plist,p)
  }
  
  box_list <- lapply(c(plist),FUN=function(x){gwasPxG(i,x)})
  #H<-do.call("grid.arrange", c(box_list, ncol=length(plist)))
  grid.draw(grid.arrange(do.call("grid.arrange", c(box_list, ncol=length(plist))),blank_plot))
  #J<-grid.arrange(H,blank_plot) #bind with blank plot to get appropriate height
  #grid.draw(J)
  
  
  
  #medians<-filter(Amedian_df,trait==i)
  # median_diff<-medians$median_diff
  #diff<-medians$diff
  #print(median_diff)
  #print(diff)
  
  
  cat("\n\n\\pagebreak\n")
}









```

